install.ts

```diff
@@ -0,0 +1,202 @@
+}
+
+async function follow(url: string): Promise<any> {
+  let located: boolean = false;
+  let res: any; // TODO: annotate deno Response
+  let count: number = 0;
+  while (!located) {
+    if (++count === 4) throw Error(`unable to fetch from ${url}`);
+    res = await fetch(url);
+    if (res.status >= 300 && res.status < 400)
+      url = res.headers.get("Location");
+    else if (res.status === 200) located = true;
+  }
+  return res;
+}
+
+async function release_url(tag?: string): Promise<{ [key: string]: string }> {
+  const url: string = tag ? `${TAG_RELEASE_URL}/${tag}` : LATEST_RELEASE_URL;
+  let filename: string;
+  switch (platform.os) {
+    case "linux":
+      filename = LINUX_GZIP;
+      break;
+    case "mac":
+      filename = OSX_GZIP;
+      break;
+    case "win":
+      filename = WIN_ZIP;
+      break;
+    default:
+      throw Error(`unsupported OS ${platform.os}`);
+  }
+  const res: any = await follow(url); // TODO: annotate deno Response
+  const link: string = (await res.text())
+    .split(/\r?\n/)
+    .find((line: string) => line.includes("href") && line.includes(filename));
+  if (!link) panic(Error(`unable to find ${filename} @ ${url}`));
+  const match: string = link.replace(/^.*href=(?:"|')([^"']*)(?:"|').*$/, "$1");
+  if (!/^\/denoland/.test(match))
+    panic(Error(`unable to find ${filename} @ ${url}`));
+  return {
+    url: `https://github.com${match}`,
+    tag: match.replace(/^.*(v\d+\.\d+\.\d+).*$/, "$1")
+  };
+}
+
+async function temp_download(
+  temp_dir: string,
+  url: string,
+  suffix: string
+): Promise<string> {
+  const res: any = await follow(url); // TODO: annotate deno Response
+  const temp_file: string = join(temp_dir, `${Date.now()}.${suffix}`);
+  await writeFile(temp_file, new Uint8Array(await res.arrayBuffer()));
+  return temp_file;
+}
+
+async function unpack_bin(archive: string): Promise<void> {
+  await mkdirp(DENO_BIN_DIR);
+  let args: string[];
+  if (WIN32) {
+    await rename(DENO_BIN, OLD_DENO_BIN);
+    args = [
+      "powershell.exe",
+      "-Command",
+      `Expand-Archive "${archive}" -DestinationPath "${DENO_BIN_DIR}"`
+    ];
+  } else {
+    args = ["gunzip", "-d", archive];
+  }
+  const child: Process = run({ args });
+  const child_status: ProcessStatus = await child.status();
+  if (!child_status.success)
+    panic(Error(`(g)unzip failed. ${args} -> ${child_status.code}`));
+  child.close();
+  if (!WIN32) {
+    if (platform.os === "linux") await rename(DENO_BIN, OLD_DENO_BIN);
+    await copyFile(archive.replace(/\.gz$/, ""), DENO_BIN);
+  }
+}
+
+async function make_handy(): Promise<void> {
+  if (WIN32) {
+    if (!proc_env.Path.includes(DENO_BIN_DIR)) {
+      const upd_path: string = `${proc_env.Path};${DENO_BIN_DIR}`;
+      const args: string[] = [
+        "powershell.exe",
+        "-Command",
+        `[Environment]::SetEnvironmentVariable("PATH","${upd_path}",` +
+          `[EnvironmentVariableTarget]::User)`
+      ];
```

Comment:
In Windows, there's two ways to modify env vars: `setx` modifies an env var for all future terminal sessions but doesn't change the value for the current session. `set` modifies an env var for the current terminal session but doesn't persist it for future sessions.

The PowerShell command above is basically only running `setx`, setting the value for future sessions. So `deno` won't be available on the Path of the current terminal session, and users would have to restart the terminal first.

This is why the `install.ps1` script is running `$env:PATH += ";${deno_dir}"` after `[Environment]::SetEnvironmentVariable`, by the way. However, since PowerShell is executed as a (child?) process here, I don't think that's gonna work (but I can't test it since I'm not on a Windows machine at the moment).

@chiefbiiko Can you look into this perhaps? :)

Reply:
I looked into possible options but could not make it work without any manual steps.
However, whenever `./install.ts` is run, deno is already installed, and the `PATH` should be set accordingly.
In the case that the binary is in place but the `PATH` is not set, `install.ts` takes care of updating the `PATH` (manual shell restart required).
As the latter should be rather unusual, informing the user about a required shell restart is okay in my opinion.

@MarkTiedemann What do you think?

Reply:
Haha, true, for a second there I forgot we're installing deno from an existing installation! I think it's fine as it is then. :)

Reply:
I even think the "start a fresh shell session" message is unnecessary in this case.